File: visual-model-ui/src/main/java/org/visual/model/ui/widget/Tray.java
package org.visual.model.ui.widget;

import javafx.application.Platform;
import javafx.stage.Stage;
import lombok.SneakyThrows;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.awt.event.ActionListener;
import java.nio.file.Path;
import java.util.Collection;

public class Tray {
    private final SystemTray tray = SystemTray.getSystemTray();
    private final Image image;
    private final PopupMenu popup;

    @SneakyThrows
    public Tray(@NotNull Path iconPath, Stage primaryStage, @NotNull Collection<MenuItem> menuItems) {
        this.image = Toolkit.getDefaultToolkit().getImage(iconPath.toAbsolutePath().toString());
        this.popup = new PopupMenu();
        menuItems.forEach(popup::add);

        ActionListener trayListener = e -> {
            if (primaryStage.isShowing()) {
                Platform.runLater(primaryStage::hide);
            } else {
                Platform.runLater(primaryStage::show);
            }
        };

        TrayIcon trayIcon = new TrayIcon(image, "Tray Demo", popup);
        trayIcon.addActionListener(trayListener);

        Platform.setImplicitExit(false);
        tray.add(trayIcon);
    }
}

File: visual-model-ui/src/main/kotlin/org/visual/model/ui/widget/Tray.kt
package org.visual.model.ui.widget

import javafx.application.Platform
import javafx.stage.Stage
import java.awt.*
import java.awt.event.ActionListener
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.io.path.absolutePathString


class Tray @JvmOverloads constructor(
    iconPath: Path = Paths.get(""),
    primaryStage: Stage,
    menuItems: Collection<MenuItem> = listOf(),
) {
    private val tray: SystemTray = SystemTray.getSystemTray()
    private val image: Image = Toolkit.getDefaultToolkit().getImage(iconPath.absolutePathString())
    private val popup: PopupMenu = PopupMenu()
        .apply {
            menuItems.forEach {
                add(it)
            }
        }

    private val trayListener: ActionListener = ActionListener {
        if (primaryStage.isShowing) Platform.runLater { primaryStage.hide() } else Platform.runLater { primaryStage.show() }
    }

    // construct a TrayIcon
    private val trayIcon = TrayIcon(image, "Tray Demo", popup).apply {
        addActionListener(trayListener)
    }

    init {
        Platform.setImplicitExit(false);
    }

    init {
        tray.add(trayIcon)
    }
}